File: packages/pq-key-encoder/ts/src/utils/validation.ts
import type { AlgorithmInfo } from 'pq-oid';
import { Algorithm, OID } from 'pq-oid';
import { InvalidInputError, KeySizeMismatchError, UnsupportedAlgorithmError } from '../errors';
import type { AlgorithmName, KeyData, KeyType } from '../types';

/** Guard that a value is a Uint8Array. */
export function assertUint8Array(value: unknown, name = 'value'): asserts value is Uint8Array {
  if (!(value instanceof Uint8Array)) {
    throw new InvalidInputError(`${name} must be a Uint8Array.`);
  }
}

/** Guard that a byte array is non-empty. */
export function assertNonEmpty(value: Uint8Array, name = 'value'): void {
  if (value.length === 0) {
    throw new InvalidInputError(`${name} must not be empty.`);
  }
}

/** Resolve algorithm metadata from the registry. */
export function getAlgorithmInfo(algorithm: AlgorithmName): AlgorithmInfo {
  try {
    return Algorithm.get(algorithm);
  } catch (error) {
    if (error instanceof Error) {
      throw new UnsupportedAlgorithmError(error.message);
    }
    throw new UnsupportedAlgorithmError('Unknown algorithm.');
  }
}

/** Convert an OID string to a known algorithm name. */
export function detectAlgorithmFromOid(oid: string): AlgorithmName {
  try {
    return OID.toName(oid) as AlgorithmName;
  } catch (error) {
    if (error instanceof Error) {
      throw new UnsupportedAlgorithmError(error.message);
    }
    throw new UnsupportedAlgorithmError('Unknown algorithm.');
  }
}

/** Return the expected key size for an algorithm and key type. */
export function expectedKeySize(algorithm: AlgorithmName, keyType: KeyType): number {
  const info = getAlgorithmInfo(algorithm);
  return keyType === 'public' ? info.publicKeySize : info.privateKeySize;
}

/** Validate that a key length matches the algorithm expectation. */
export function validateKeySize(
  algorithm: AlgorithmName,
  keyType: KeyType,
  bytes: Uint8Array,
): void {
  const expected = expectedKeySize(algorithm, keyType);
  if (bytes.length !== expected) {
    throw new KeySizeMismatchError(algorithm, keyType, expected, bytes.length);
  }
}

/** Validate key data and optionally enforce a specific key type. */
export function assertKeyData(key: KeyData, expectedType?: KeyType): void {
  if (expectedType && key.type !== expectedType) {
    throw new InvalidInputError(`Expected a ${expectedType} key.`);
  }
  assertUint8Array(key.bytes, 'key.bytes');
  assertNonEmpty(key.bytes, 'key.bytes');
  validateKeySize(key.alg, key.type, key.bytes);
}

File: packages/pq-key-encoder/rust/src/validation.rs
use pq_oid::Algorithm;

use crate::error::{Error, Result};
use crate::types::KeyType;

/// Returns the expected key size in bytes for the given algorithm and key type.
pub(crate) fn expected_key_size(algorithm: Algorithm, key_type: KeyType) -> usize {
    match key_type {
        KeyType::Public => algorithm.public_key_size(),
        KeyType::Private => algorithm.private_key_size(),
    }
}

/// Validates that the key bytes have the correct length for the algorithm and key type.
pub(crate) fn validate_key_size(
    algorithm: Algorithm,
    key_type: KeyType,
    bytes: &[u8],
) -> Result<()> {
    if bytes.is_empty() {
        return Err(Error::EmptyKey);
    }
    let expected = expected_key_size(algorithm, key_type);
    if bytes.len() != expected {
        return Err(Error::KeySizeMismatch {
            algorithm,
            key_type,
            expected,
            actual: bytes.len(),
        });
    }
    Ok(())
}
